rfcs/2021-04-13-7027-core-extraction.md

```diff
@@ -0,0 +1,319 @@
+* resolving any incompatibilities between their new code and the existing vector
+  model and
+* waiting for CI to approve their changes.
+
+With regard to benchmarks we generally do not request contributors to add them
+as doing so is too complicated. Our flat structure promotes a conflation between
+unit and integrated benchmark tests, not to mention you need a fair bit of
+memory to _link_ vector to run these tests yourself. The feedback loop here is
+hours long. Any incompatibilities between new and existing code are quite
+time-consuming to resolve as our flat structure encourages a "big bang" mindset
+to changes, if not mandates it. The flatness of our project structure does mean,
+however, that relatively simple sources/transforms/sinks are convenient to
+introduce, there is "one place" for such things.
+
+The vector project does have some sub-packages. VRL -- in `lib/vrl` -- is one
+such. The development experience here has a tighter feedback loop than in vector
+proper. VRL benchmarks can be built independent of vector -- avoiding its link
+cost -- and its tests end up clearly focused on VRL itself. This _does_ mean
+that integration tests remain to be written in the top-level of the project but
+the overall experience is improved. Changes in VRL are isolated to VRL, though
+experience with `Lookup`/`LookupBuf` -- see
+[5374](https://github.com/timberio/vector/pull/5374) -- does suggest that
+aggressive integration of a cross-cutting concern will make things challenging.
+
+We hope to capture the benefits to development we've seen in the VRL sub-package
+and make them more broadly available in the vector project.
+
+## The Structure of the Project
+
+Today there are 168 top-level features in the project, as measured on commit
+48d2a84b1b11ba54db7bd892944f2a479238edb4:
+
+```sh
+> cargo read-manifest | jq ".features" | jq 'keys' | wc -l
+[
+  "all-integration-tests",
+  "all-logs",
+  "all-metrics",
+  "api",
+  "api-client",
+  "aws-cloudwatch-logs-integration-tests",
+  "aws-cloudwatch-metrics-integration-tests",
+  "aws-ec2-metadata-integration-tests",
+  "aws-ecs-metrics-integration-tests",
+  "aws-integration-tests",
+  "aws-kinesis-firehose-integration-tests",
+  "aws-kinesis-streams-integration-tests",
+  "aws-s3-integration-tests",
+  "aws-sqs-integration-tests",
+  "benches",
+  "cli-tests",
+  "clickhouse-integration-tests",
+  "default",
+  "default-cmake",
+  "default-msvc",
+  "default-musl",
+  "default-no-api-client",
+  "default-no-vrl-cli",
+  "disable-resolv-conf",
+  "docker-logs-integration-tests",
+  "es-integration-tests",
+  "gcp-cloud-storage-integration-tests",
+  "gcp-integration-tests",
+  "gcp-pubsub-integration-tests",
+  "humio-integration-tests",
+  "influxdb-integration-tests",
+  "kafka-integration-tests",
+  "kubernetes",
+<SNIP>
+```
+
+The features are split in their purposes. Some flag on tests, some are for build
+targets, some map -- not quite precisely -- to source/transform/sinks being
```

Comment:
I am curious in what sense you find this mapping to be not quite precise.

Reply:
When I was poking around Cargo.toml recently in #7076 I found features that didn't flag on dependencies, or assumed, implicitly that another feature would be flagged on. I think these are corrected. I'll remove this parenthetical.